### PR TITLE
Add customizable images from config

### PR DESCRIPTION
**Pull request by misohu**
- Add custom_images config option to populate seldon configmap images
- Add integration tests

### CODE REVIEW

**Review by DnPlas — COMMENTED** [review]
Thanks @misohu ! Some comments.

File: tests/integration/test_charm.py [hunk 1]
```diff
@@ -56,6 +69,55 @@ async def test_build_and_deploy(ops_test: OpsTest):
...
+async def test_blocked_on_invalid_config(ops_test: OpsTest):
+    await ops_test.model.applications[APP_NAME].set_config({"custom_images": "{"})
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME], status="blocked", raise_on_blocked=False, timeout=300
+    )
+    assert ops_test.model.applications[APP_NAME].units[0].workload_status == "blocked"
+
+
+@pytest.mark.abort_on_fail
+async def test_back_to_active_on_config(lightkube_client: Client, ops_test: OpsTest):
```
Comment (DnPlas) [inline tests/integration/test_charm.py:109]: This test case is alright, but I'd change the name to ` test_back_to_active_on_default_config`, as we want a working seldon-core-operator and its activators to be active and functional before we proceed with a bunch of other test cases.

File: tests/integration/test_charm.py [hunk 2]
```diff
@@ -56,6 +69,55 @@ async def test_build_and_deploy(ops_test: OpsTest):
     assert ops_test.model.applications[APP_NAME].units[0].workload_status == "active"
 
 
+@pytest.mark.abort_on_fail
+async def test_configmap_created(lightkube_client: Client, ops_test: OpsTest):
```
Comment (DnPlas) [inline tests/integration/test_charm.py:73]: I think this test case can live in unit test, but if it is too much of a problem, I'm okay with it.

File: src/charm.py [hunk 3]
```diff
@@ -251,6 +290,52 @@ def _seldon_core_operator_layer(self) -> Layer:
...
+        images = default_images
+        for image_name, custom_image in custom_images.items():
+            if custom_image:
+                if image_name in images:
+                    images[image_name] = custom_image
+                else:
+                    self.logger.warning(f"image_name {image_name} not in image list, ignoring.")
+
+        # This are special cases comfigmap where they need to be split into image and version
+        for image_name in [
```
Comment (i-chvets) [inline src/charm.py:321]: Can we add a comment here that this list needs to be in-sync with `configmap.yaml.j2`?
I.e. if that file is updated with new images, for example, we will revise this list.

File: tests/integration/test_data/expected_changed_seldon_cm.json
[the remark this answers is not in the document]
Comment (misohu) [reply tests/integration/test_data/expected_changed_seldon_cm.json:1]: The values there are strings I am sorry :(

File: tests/integration/test_charm.py [hunk 4]
```diff
@@ -56,6 +69,55 @@ async def test_build_and_deploy(ops_test: OpsTest):
...
+    )
+    seldon_config_cm = lightkube_client.get(
+        ConfigMap, SELDON_CM_NAME, namespace=ops_test.model_name
+    )
+
+    assert seldon_config_cm.data == SELDON_CONFIG_CHANGED
+
+
+@pytest.mark.abort_on_fail
+async def test_blocked_on_invalid_config(ops_test: OpsTest):
```
[the remark this answers is not in the document]
Comment (misohu) [reply tests/integration/test_charm.py:100]: Added unit test based on Andrews PR but I will keep also integration test as they test that the actual configmap exists :)

**Comment by i-chvets** [comment]
LGTM. Let's resovle pending conversations.

**Review by i-chvets — DISMISSED** [review]
Approved.

**Review by DnPlas — COMMENTED** [review]
Just few more comments, but looking good already!

File: tests/integration/test_charm.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (misohu) [reply tests/integration/test_charm.py]: After discussion I removed the integration test.

File: README.md [hunk 5]
```diff
@@ -67,6 +67,18 @@ Please see the [official docs site](https://charmed-kubeflow.io/docs) for comple
 If you find a bug in our operator or want to request a specific feature, please file a bug here: 
 [https://github.com/canonical/seldon-core-operator/issues](https://github.com/canonical/seldon-core-operator/issues)
 
+### Setting Custom Images for Knative Serving
+
+Seldon deploys with a set of preconfigured images.  These images, listed in the `src/default-custom-images.json`, can be overridden using the charm config `custom_images`.
```
[the remark this answers is not in the document]
Comment (misohu) [reply README.md:72]: There is no upstream documentation link like the one in knative.

**Review by i-chvets — APPROVED** [review]
Approved.

**Comment by i-chvets** [comment]
@misohu Commits need to be signed for this repo, otherwise we won't be able to merge this PR into main.

**Comment by misohu** [comment]
Signed commits in this PR @i-chvets https://github.com/canonical/seldon-core-operator/pull/198 thanks

**Pull request closed without merging**